planner/core/exhaust_physical_plans.go

```diff
@@ -1612,6 +1647,74 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
 	return joins, true
 }
 
+func (p *LogicalJoin) tryToGetMppHashJoin(prop *property.PhysicalProperty, useBCJ bool) []PhysicalPlan {
+	if !prop.IsEmpty() {
+		return nil
+	}
+	if prop.TaskTp != property.RootTaskType && prop.TaskTp != property.MppTaskType {
+		return nil
+	}
+
+	lkeys, rkeys, _, _ := p.GetJoinKeys()
+	// check match property
+	baseJoin := basePhysicalJoin{
+		JoinType:        p.JoinType,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		OtherConditions: p.OtherConditions,
+		DefaultValues:   p.DefaultValues,
+		LeftJoinKeys:    lkeys,
+		RightJoinKeys:   rkeys,
+	}
+
+	preferredBuildIndex := 0
+	if p.JoinType == InnerJoin {
+		if p.children[0].statsInfo().Count() > p.children[1].statsInfo().Count() {
+			preferredBuildIndex = 1
+		}
+	} else if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin || p.JoinType == LeftOuterJoin {
+		preferredBuildIndex = 1
+	}
+	baseJoin.InnerChildIdx = preferredBuildIndex
+	childrenProps := make([]*property.PhysicalProperty, 2)
+	if useBCJ {
+		childrenProps[preferredBuildIndex] = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.BroadcastType, Enforced: true}
+		expCnt := math.MaxFloat64
+		if prop.ExpectedCnt < p.stats.RowCount {
+			expCntScale := prop.ExpectedCnt / p.stats.RowCount
+			expCnt = p.children[1-preferredBuildIndex].statsInfo().RowCount * expCntScale
+		}
+		if prop.PartitionTp == property.HashType {
+			hashKeys := rkeys
+			if preferredBuildIndex == 0 {
+				hashKeys = lkeys
+			}
+			if prop.MatchPartCols(hashKeys) {
+				childrenProps[1-preferredBuildIndex] = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: expCnt, PartitionTp: property.HashType, PartitionCols: hashKeys}
+			} else {
+				return nil
+			}
+		} else {
+			childrenProps[1-preferredBuildIndex] = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: expCnt, PartitionTp: property.AnyType}
+		}
+	} else {
+		if prop.PartitionTp == property.HashType {
+			if !prop.MatchPartCols(lkeys) && !prop.MatchPartCols(rkeys) {
+				return nil
+			}
+		}
+		childrenProps[0] = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: lkeys, Enforced: true}
+		childrenProps[1] = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: rkeys, Enforced: true}
```

Comment:
need to rewrite here
1. how to process three other cases for `if !prop.MatchPartCols(lkeys) && !prop.MatchPartCols(rkeys)`?
2. `PartitionCols ` should be the prop of its parpent? i.e., how to process the relationship between its join keys, and parent's partitionCols?

Reply:
1. for other cases just enforce it.
2. I don't know your meanings, can you detail it ?

Reply:
1. it would need to enforce one side , even neither of each side.
2. partition keys are different from join keys.

Reply:
these codes should be more complex than those of the broadcast join. It is the key to the performance of multi-table joins.

Reply:
1. yeah, this code can process this situation. enforce = true doesn't mean it's sure to exist an exchanger. It will check whether the keys match.
2. yeah, but it's really complex and let's leave it to future.

Reply:
add some TODO to refine it later after refer to some related work.